**Coursebook: Data Wrangling and Group by Aggregation**
- Part 3 of Data Analytics Specialization
- Course Length: 12 hours
- Last Updated: April 2019
___

- Author: [Samuel Chan](https://github.com/onlyphantom)
- Developed by [Algoritma](https://algorit.ma)'s product division and instructors team

# Background

## Top-Down Approach 

The coursebook is part of the **Data Analytics Specialization** offered by [Algoritma](https://algorit.ma). It takes a more accessible approach compared to Algoritma's core educational products, by getting participants to overcome the "how" barrier first, rather than a detailed breakdown of the "why". 

This translates to an overall easier learning curve, one where the reader is prompted to write short snippets of code in frequent intervals, before being offered an explanation on the underlying theoretical frameworks. Instead of mastering the syntactic design of the Python programming language, then moving into data structures, and then the `pandas` library, and then the mathematical details in an imputation algorithm, and its code implementation; we would do the opposite: Implement the imputation, then a succinct explanation of why it works and applicational considerations (what to look out for, what are assumptions it made, when _not_ to use it etc).

## Learn-by-Building

This coursebook is intended for participants who have completed the preceding courses offered in the **Data Analytics Developer Specialization**. This is the third course, **Data Wrangling and Group By Aggregation**.

The coursebook focuses on:
- Stacking and Unstacking
- Working with MultiIndex DataFrames
- Reshaping your DataFrame with Melt
- Using Group By Effectively

At the end of this course is a Graded Asssignment section, where you are expected to apply all that you've learned on a new dataset, and attempt the given questions.

# Data Wrangling and Reshaping

In the previous two courses, we've got our hands on a few common techniques and learned how to explore data using `pandas` built-in methods. Specifically, we've  in the first and second part of this series how to use the following inspection, diagnostic and exploratory tools: 

**Data Inspection**
- `.head()` and `.tail()`
- `.describe()`
- `.shape` and `.size`
- `.axes`
- `.dtypes`
- Subsetting using `.loc`, `.iloc` and conditionals
---
**Diagnostic and Exploratory**
- Tables
- Cross-Tables and Aggregates
- Using `aggfunc` for aggregate functions
- Pivot Tables
- Working with DateTime
- Working with Categorical Data
- Duplicates and Missing Value Treatment

The first half of this course serves as an extension from the last. We'll pick up some new techniques to supplement our EDA toolset. Let us begin with reshaping techniques. 

In [32]:
from pandas_datareader import data
import pandas as pd

In [33]:
symbol = ['AAPL', 'FB', 'GOOGL']
source = 'yahoo'
start_date = '2018-01-01'
end_date = '2019-04-24'
stock = data.DataReader(symbol, source, start_date, end_date)
stock.head()

Attributes        High                                  Low              \
Symbols           AAPL          FB        GOOGL        AAPL          FB   
Date                                                                      
2018-01-02  172.300003  181.580002  1075.979980  169.259995  177.550003   
2018-01-03  174.550003  184.779999  1096.099976  171.960007  181.330002   
2018-01-04  173.470001  186.210007  1104.079956  172.080002  184.100006   
2018-01-05  175.369995  186.899994  1113.579956  173.050003  184.929993   
2018-01-08  175.610001  188.899994  1119.160034  173.929993  186.330002   

Attributes                     Open                                Close  \
Symbols           GOOGL        AAPL          FB        GOOGL        AAPL   
Date                                                                       
2018-01-02  1053.020020  170.160004  177.679993  1053.020020  172.259995   
2018-01-03  1073.430054  172.529999  181.880005  1073.930054  172.229996   
2018-01-04  1094.260010  172.539993  184.899994  1097.089966  173.029999   
2018-01-05  1101.800049  173.440002  185.589996  1103.449951  175.000000   
2018-01-08  1110.000000  174.350006  187.199997  1111.000000  174.350006   

Attributes                               Volume                         \
Symbols             FB        GOOGL        AAPL          FB      GOOGL   
Date                                                                     
2018-01-02  181.419998  1073.209961  25555900.0  18151900.0  1588300.0   
2018-01-03  184.669998  1091.520020  29517900.0  16886600.0  1565900.0   
2018-01-04  184.330002  1095.760010  22434600.0  13880900.0  1302600.0   
2018-01-05  186.850006  1110.290039  23660000.0  13574500.0  1512500.0   
2018-01-08  188.279999  1114.209961  20567800.0  17994700.0  1232200.0   

Attributes   Adj Close                           
Symbols           AAPL          FB        GOOGL  
Date                                             
2018-01-02  168.339050  181.419998  1073.209961  
2018-01-03  168.309738  184.669998  1091.520020  
2018-01-04  169.091522  184.330002  1095.760010  
2018-01-05  171.016678  186.850006  1110.290039  
2018-01-08  170.381485  188.279999  1114.209961

If you do not have the `pandas_datareader` module installed, or if you're following along this coursebook without an active connection, you can instead load it from the serialized object I stored in your `data_cache` folder. 

Creating the DataFrame object by reading from `pickle`:
- `stock = pd.read_pickle('data_cache/stock')`

Serializing the DataFrame object to a byte stream using `pickle`:
- `stock.to_pickle('data_cache/stock')`

In [34]:
stock.to_pickle('data_cache/stock')
stock = pd.read_pickle('data_cache/stock')
stock.head()

Attributes        High                                  Low              \
Symbols           AAPL          FB        GOOGL        AAPL          FB   
Date                                                                      
2018-01-02  172.300003  181.580002  1075.979980  169.259995  177.550003   
2018-01-03  174.550003  184.779999  1096.099976  171.960007  181.330002   
2018-01-04  173.470001  186.210007  1104.079956  172.080002  184.100006   
2018-01-05  175.369995  186.899994  1113.579956  173.050003  184.929993   
2018-01-08  175.610001  188.899994  1119.160034  173.929993  186.330002   

Attributes                     Open                                Close  \
Symbols           GOOGL        AAPL          FB        GOOGL        AAPL   
Date                                                                       
2018-01-02  1053.020020  170.160004  177.679993  1053.020020  172.259995   
2018-01-03  1073.430054  172.529999  181.880005  1073.930054  172.229996   
2018-01-04  1094.260010  172.539993  184.899994  1097.089966  173.029999   
2018-01-05  1101.800049  173.440002  185.589996  1103.449951  175.000000   
2018-01-08  1110.000000  174.350006  187.199997  1111.000000  174.350006   

Attributes                               Volume                         \
Symbols             FB        GOOGL        AAPL          FB      GOOGL   
Date                                                                     
2018-01-02  181.419998  1073.209961  25555900.0  18151900.0  1588300.0   
2018-01-03  184.669998  1091.520020  29517900.0  16886600.0  1565900.0   
2018-01-04  184.330002  1095.760010  22434600.0  13880900.0  1302600.0   
2018-01-05  186.850006  1110.290039  23660000.0  13574500.0  1512500.0   
2018-01-08  188.279999  1114.209961  20567800.0  17994700.0  1232200.0   

Attributes   Adj Close                           
Symbols           AAPL          FB        GOOGL  
Date                                             
2018-01-02  168.339050  181.419998  1073.209961  
2018-01-03  168.309738  184.669998  1091.520020  
2018-01-04  169.091522  184.330002  1095.760010  
2018-01-05  171.016678  186.850006  1110.290039  
2018-01-08  170.381485  188.279999  1114.209961

Create a DataFrame by subsetting only the `Close` columns. Name it `closingprice`. Then, use `.isna().sum()` to count the number of missing values in each of the columns present in `closingprice`.

If there are any missing values, use the `.fillna(method='ffill')` method to fill those missing values:

In [35]:
## Your code below
closingprice = stock.loc[:, ["Close"]]
closingprice.head()
## -- Solution code

Attributes       Close                         
Symbols           AAPL          FB        GOOGL
Date                                           
2018-01-02  172.259995  181.419998  1073.209961
2018-01-03  172.229996  184.669998  1091.520020
2018-01-04  173.029999  184.330002  1095.760010
2018-01-05  175.000000  186.850006  1110.290039
2018-01-08  174.350006  188.279999  1114.209961

In [36]:
closingprice.isna().count()

Attributes  Symbols
Close       AAPL       329
            FB         329
            GOOGL      329
dtype: int64

In [37]:
closingprice = closingprice.fillna(method='ffill')
closingprice.head()

Attributes       Close                         
Symbols           AAPL          FB        GOOGL
Date                                           
2018-01-02  172.259995  181.419998  1073.209961
2018-01-03  172.229996  184.669998  1091.520020
2018-01-04  173.029999  184.330002  1095.760010
2018-01-05  175.000000  186.850006  1110.290039
2018-01-08  174.350006  188.279999  1114.209961

In [38]:
closingprice.isna().count()

Attributes  Symbols
Close       AAPL       329
            FB         329
            GOOGL      329
dtype: int64

If you pay close attention to the index of `stock`, you may already realized by now that there are days where no records were present. 2018-01-01, 2018-01-06, and 2018-01-07 were absent from our DataFrame because they happen to fall on weekends.

While the trading hours of [different stock markets differ](https://www.maybank-ke.com.sg/markets/markets-listing/trading-hours/) (the NYSE for example open its market floor from 9.30am to 4pm five days a week), on weekends as well as federal holidays all stock exchanges are closed for business.

We can create (or recreate) the index by passing in our own values. In the following cell we created a date range and create the index using that new date range:

In [39]:
closingprice = stock['Close']
quarter1 = pd.date_range(start="2019-01-01", end="2019-03-31")
closingprice = closingprice.reindex(quarter1)
closingprice.head(8)

Symbols           AAPL          FB        GOOGL
2019-01-01         NaN         NaN          NaN
2019-01-02  157.919998  135.679993  1054.680054
2019-01-03  142.190002  131.740005  1025.469971
2019-01-04  148.259995  137.949997  1078.069946
2019-01-05         NaN         NaN          NaN
2019-01-06         NaN         NaN          NaN
2019-01-07  147.929993  138.050003  1075.920044
2019-01-08  150.750000  142.529999  1085.369995

Now use forward-fill to fill the `NA` values:

In [40]:
## Your code below
closingprice = closingprice.fillna(method="ffill")
closingprice.head(8)
## -- Solution code

Symbols           AAPL          FB        GOOGL
2019-01-01         NaN         NaN          NaN
2019-01-02  157.919998  135.679993  1054.680054
2019-01-03  142.190002  131.740005  1025.469971
2019-01-04  148.259995  137.949997  1078.069946
2019-01-05  148.259995  137.949997  1078.069946
2019-01-06  148.259995  137.949997  1078.069946
2019-01-07  147.929993  138.050003  1075.920044
2019-01-08  150.750000  142.529999  1085.369995

## `stack()` and `unstack()`

`stack()` stack the prescribed level(s) from columns to index and is particularly useful on DataFrames having a multi-level columns. It does so by "shifting" the columns to create new levels on its index. 

This is easier understood when we just see an example. Notice that `stock` has a 2-level column (Attributes and Symbols) and 1-level index (Date):

In [41]:
stock.head(8)

Attributes        High                                  Low              \
Symbols           AAPL          FB        GOOGL        AAPL          FB   
Date                                                                      
2018-01-02  172.300003  181.580002  1075.979980  169.259995  177.550003   
2018-01-03  174.550003  184.779999  1096.099976  171.960007  181.330002   
2018-01-04  173.470001  186.210007  1104.079956  172.080002  184.100006   
2018-01-05  175.369995  186.899994  1113.579956  173.050003  184.929993   
2018-01-08  175.610001  188.899994  1119.160034  173.929993  186.330002   
2018-01-09  175.059998  188.800003  1118.439941  173.410004  187.100006   
2018-01-10  174.300003  187.889999  1112.780029  173.000000  185.630005   
2018-01-11  175.490005  188.399994  1114.849976  174.490005  187.380005   

Attributes                     Open                                Close  \
Symbols           GOOGL        AAPL          FB        GOOGL        AAPL   
Date                                                                       
2018-01-02  1053.020020  170.160004  177.679993  1053.020020  172.259995   
2018-01-03  1073.430054  172.529999  181.880005  1073.930054  172.229996   
2018-01-04  1094.260010  172.539993  184.899994  1097.089966  173.029999   
2018-01-05  1101.800049  173.440002  185.589996  1103.449951  175.000000   
2018-01-08  1110.000000  174.350006  187.199997  1111.000000  174.350006   
2018-01-09  1108.199951  174.550003  188.699997  1118.439941  174.330002   
2018-01-10  1103.979980  173.160004  186.940002  1107.000000  174.289993   
2018-01-11  1106.479980  174.589996  188.399994  1112.310059  175.279999   

Attributes                               Volume                         \
Symbols             FB        GOOGL        AAPL          FB      GOOGL   
Date                                                                     
2018-01-02  181.419998  1073.209961  25555900.0  18151900.0  1588300.0   
2018-01-03  184.669998  1091.520020  29517900.0  16886600.0  1565900.0   
2018-01-04  184.330002  1095.760010  22434600.0  13880900.0  1302600.0   
2018-01-05  186.850006  1110.290039  23660000.0  13574500.0  1512500.0   
2018-01-08  188.279999  1114.209961  20567800.0  17994700.0  1232200.0   
2018-01-09  187.869995  1112.790039  21584000.0  12393100.0  1340400.0   
2018-01-10  187.839996  1110.140015  23959900.0  10529900.0  1036700.0   
2018-01-11  187.770004  1112.050049  18667700.0   9588600.0  1121200.0   

Attributes   Adj Close                           
Symbols           AAPL          FB        GOOGL  
Date                                             
2018-01-02  168.339050  181.419998  1073.209961  
2018-01-03  168.309738  184.669998  1091.520020  
2018-01-04  169.091522  184.330002  1095.760010  
2018-01-05  171.016678  186.850006  1110.290039  
2018-01-08  170.381485  188.279999  1114.209961  
2018-01-09  170.361954  187.869995  1112.790039  
2018-01-10  170.322845  187.839996  1110.140015  
2018-01-11  171.290329  187.770004  1112.050049

When we stack the `stock` DataFrame, we shrink the number of levels on its column by one: `stock` now has 1-level column named `Attributes`: 

In [42]:
stock.stack().head(8)

Attributes                 High          Low         Open        Close  \
Date       Symbols                                                       
2018-01-02 AAPL      172.300003   169.259995   170.160004   172.259995   
           FB        181.580002   177.550003   177.679993   181.419998   
           GOOGL    1075.979980  1053.020020  1053.020020  1073.209961   
2018-01-03 AAPL      174.550003   171.960007   172.529999   172.229996   
           FB        184.779999   181.330002   181.880005   184.669998   
           GOOGL    1096.099976  1073.430054  1073.930054  1091.520020   
2018-01-04 AAPL      173.470001   172.080002   172.539993   173.029999   
           FB        186.210007   184.100006   184.899994   184.330002   

Attributes              Volume    Adj Close  
Date       Symbols                           
2018-01-02 AAPL     25555900.0   168.339050  
           FB       18151900.0   181.419998  
           GOOGL     1588300.0  1073.209961  
2018-01-03 AAPL     29517900.0   168.309738  
           FB       16886600.0   184.669998  
           GOOGL     1565900.0  1091.520020  
2018-01-04 AAPL     22434600.0   169.091522  
           FB       13880900.0   184.330002

`unstack()` does the opposite: it "shifts" the levels from index axis onto column axis. Try and create a stack DataFrame, and then apply `unstack` on the new DataFrame to see it return to the original shape:

In [43]:
## Your code below

stockstack = stock.stack()
stockstack.head(8)

## -- Solution code

Attributes                 High          Low         Open        Close  \
Date       Symbols                                                       
2018-01-02 AAPL      172.300003   169.259995   170.160004   172.259995   
           FB        181.580002   177.550003   177.679993   181.419998   
           GOOGL    1075.979980  1053.020020  1053.020020  1073.209961   
2018-01-03 AAPL      174.550003   171.960007   172.529999   172.229996   
           FB        184.779999   181.330002   181.880005   184.669998   
           GOOGL    1096.099976  1073.430054  1073.930054  1091.520020   
2018-01-04 AAPL      173.470001   172.080002   172.539993   173.029999   
           FB        186.210007   184.100006   184.899994   184.330002   

Attributes              Volume    Adj Close  
Date       Symbols                           
2018-01-02 AAPL     25555900.0   168.339050  
           FB       18151900.0   181.419998  
           GOOGL     1588300.0  1073.209961  
2018-01-03 AAPL     29517900.0   168.309738  
           FB       16886600.0   184.669998  
           GOOGL     1565900.0  1091.520020  
2018-01-04 AAPL     22434600.0   169.091522  
           FB       13880900.0   184.330002

In [44]:
stockstack.unstack()

Attributes        High                                  Low              \
Symbols           AAPL          FB        GOOGL        AAPL          FB   
Date                                                                      
2018-01-02  172.300003  181.580002  1075.979980  169.259995  177.550003   
2018-01-03  174.550003  184.779999  1096.099976  171.960007  181.330002   
2018-01-04  173.470001  186.210007  1104.079956  172.080002  184.100006   
2018-01-05  175.369995  186.899994  1113.579956  173.050003  184.929993   
2018-01-08  175.610001  188.899994  1119.160034  173.929993  186.330002   
2018-01-09  175.059998  188.800003  1118.439941  173.410004  187.100006   
2018-01-10  174.300003  187.889999  1112.780029  173.000000  185.630005   
2018-01-11  175.490005  188.399994  1114.849976  174.490005  187.380005   
2018-01-12  177.360001  181.479996  1131.300049  175.649994  177.399994   
2018-01-16  179.389999  181.750000  1148.880005  176.139999  178.039993   
2018-01-17  179.250000  179.320007  1139.319946  175.070007  175.800003   
2018-01-18  180.100006  180.979996  1140.589966  178.250000  177.080002   
2018-01-19  179.580002  182.369995  1143.780029  177.410004  180.169998   
2018-01-22  177.779999  185.389999  1166.880005  176.600006  180.410004   
2018-01-23  179.440002  189.550003  1178.510010  176.820007  185.550003   
2018-01-24  177.300003  190.660004  1187.050049  173.199997  186.520004   
2018-01-25  174.949997  188.619995  1185.000000  170.529999  186.600006   
2018-01-26  172.000000  190.000000  1187.560059  170.059998  186.809998   
2018-01-29  170.160004  188.839996  1198.000000  167.070007  185.630005   
2018-01-30  167.369995  188.179993  1187.930054  164.699997  181.839996   
2018-01-31  168.440002  189.830002  1186.319946  166.500000  185.220001   
2018-02-01  168.619995  195.320007  1187.449951  166.759995  187.889999   
2018-02-02  166.800003  194.210007  1131.300049  160.100006  189.979996   
2018-02-05  163.880005  190.610001  1114.989990  156.000000  180.610001   
2018-02-06  163.720001  185.770004  1087.380005  154.000000  177.740005   
2018-02-07  163.399994  185.080002  1086.530029  159.070007  179.949997   
2018-02-08  161.000000  181.839996  1063.930054  155.029999  171.479996   
2018-02-09  157.889999  176.899994  1051.719971  150.240005  167.179993   
2018-02-12  163.889999  177.550003  1065.569946  157.509995  171.839996   
2018-02-13  164.750000  175.970001  1061.219971  161.649994  173.100006   
...                ...         ...          ...         ...         ...   
2019-03-13  183.300003  174.029999  1207.119995  180.919998  172.119995   
2019-03-14  184.100006  171.149994  1204.699951  182.559998  168.160004   
2019-03-15  187.330002  167.580002  1201.719971  183.740005  162.509995   
2019-03-18  188.389999  163.899994  1194.949951  185.789993  159.279999   
2019-03-19  188.990005  163.820007  1203.410034  185.919998  160.820007   
2019-03-20  189.490005  166.119995  1229.989990  184.729996  161.240005   
2019-03-21  196.330002  166.389999  1236.439941  189.809998  163.750000   
2019-03-22  197.690002  167.419998  1233.630005  190.779999  164.089996   
2019-03-25  191.979996  166.539993  1209.439941  186.600006  162.000000   
2019-03-26  192.880005  169.449997  1207.650024  184.580002  166.350006   
2019-03-27  189.759995  168.940002  1191.920044  186.550003  164.789993   
2019-03-28  189.559998  166.720001  1177.250000  187.529999  163.330002   
2019-03-29  190.080002  167.190002  1183.969971  188.539993  164.809998   
2019-04-01  191.679993  168.899994  1200.199951  188.380005  167.279999   
2019-04-02  194.460007  174.899994  1205.810059  191.050003  169.550003   
2019-04-03  196.500000  177.960007  1220.599976  193.149994  172.949997   
2019-04-04  196.369995  178.000000  1220.550049  193.139999  175.529999   
2019-04-05  197.100006  177.000000  1220.390015  195.929993  175.100006   
2019-04-08  200.229996  175.500000  1213.949951  196.339996  174.229996   
2019-04-09  202.850006  1

### Knowledge Check

Which of the following statement is correct?

- [x] `stack()` changes the DataFrame from wide to long
- [x] `unstack()` changes the DataFrame from long to wide
- [ ] `unstack()` changes the DataFrame from wide to long

## Melt

Speaking of reshaping a DataFrame from wide format to long, another method that should be in your toolset is `melt()`. Consider the following DataFrame, which is created from `pandas` MultiIndex Slicers method, `.xs()`:

In [51]:
aapl = stock.xs('AAPL', level='Symbols', axis=1)
aapl.head()

Attributes        High         Low        Open       Close      Volume  \
Date                                                                     
2018-01-02  172.300003  169.259995  170.160004  172.259995  25555900.0   
2018-01-03  174.550003  171.960007  172.529999  172.229996  29517900.0   
2018-01-04  173.470001  172.080002  172.539993  173.029999  22434600.0   
2018-01-05  175.369995  173.050003  173.440002  175.000000  23660000.0   
2018-01-08  175.610001  173.929993  174.350006  174.350006  20567800.0   

Attributes   Adj Close  
Date                    
2018-01-02  168.339050  
2018-01-03  168.309738  
2018-01-04  169.091522  
2018-01-05  171.016678  
2018-01-08  170.381485

In [53]:
stock.xs('High', axis = 1, level = "Attributes").head()

Symbols           AAPL          FB        GOOGL
Date                                           
2018-01-02  172.300003  181.580002  1075.979980
2018-01-03  174.550003  184.779999  1096.099976
2018-01-04  173.470001  186.210007  1104.079956
2018-01-05  175.369995  186.899994  1113.579956
2018-01-08  175.610001  188.899994  1119.160034

In [54]:
aapl.shape

(329, 6)

The DataFrame above is wide: it has 329 rows and 6 columns. The `melt()` function gathers all the columns into one and store the value corresponding to each column such that the resulting DataFrame has 329 * 6 = 1,974 rows, along with the identifier and values columns:

In [57]:
aapl_melted = aapl.melt()
aapl_melted.head()

Attributes       value
0       High  172.300003
1       High  174.550003
2       High  173.470001
3       High  175.369995
4       High  175.610001

In [58]:
aapl_melted.shape

(1974, 2)

We can optionally specify one or more columns to be identifier variables (`id_vars`), which treat all other columns as value variables (`value_vars`):

In [59]:
aapl.reset_index().melt(id_vars='Date')[325:333]

Date Attributes       value
325 2019-04-18       High  204.149994
326 2019-04-22       High  204.940002
327 2019-04-23       High  207.750000
328 2019-04-24       High  208.479996
329 2018-01-02        Low  169.259995
330 2018-01-03        Low  171.960007
331 2018-01-04        Low  172.080002
332 2018-01-05        Low  173.050003

#### Knowledge Check

Add a cell below, and fill the missing values in `aapl` using any method deemed appropriate:

In [60]:
march = pd.date_range(start="2018-03-01", end="2019-03-31")
aapl = stock.xs('AAPL', level='Symbols', axis=1)
aapl = aapl.reindex(march)
aapl.head(10)

Attributes        High         Low        Open       Close      Volume  \
2018-03-01  179.779999  172.660004  178.539993  175.000000  48802000.0   
2018-03-02  176.300003  172.449997  172.800003  176.210007  38454000.0   
2018-03-03         NaN         NaN         NaN         NaN         NaN   
2018-03-04         NaN         NaN         NaN         NaN         NaN   
2018-03-05  177.740005  174.520004  175.210007  176.820007  28401400.0   
2018-03-06  178.250000  176.130005  177.910004  176.669998  23788500.0   
2018-03-07  175.850006  174.270004  174.940002  175.029999  31703500.0   
2018-03-08  177.119995  175.070007  175.479996  176.940002  23774100.0   
2018-03-09  180.000000  177.389999  177.960007  179.979996  32185200.0   
2018-03-10         NaN         NaN         NaN         NaN         NaN   

Attributes   Adj Close  
2018-03-01  171.713928  
2018-03-02  172.901230  
2018-03-03         NaN  
2018-03-04         NaN  
2018-03-05  173.499771  
2018-03-06  173.352600  
2018-03-07  171.743393  
2018-03-08  173.617508  
2018-03-09  176.600418  
2018-03-10         NaN

In [61]:
## Your code below
aapl = aapl.fillna(method="ffill")
aapl.head(10)
## -- Solution code

Attributes        High         Low        Open       Close      Volume  \
2018-03-01  179.779999  172.660004  178.539993  175.000000  48802000.0   
2018-03-02  176.300003  172.449997  172.800003  176.210007  38454000.0   
2018-03-03  176.300003  172.449997  172.800003  176.210007  38454000.0   
2018-03-04  176.300003  172.449997  172.800003  176.210007  38454000.0   
2018-03-05  177.740005  174.520004  175.210007  176.820007  28401400.0   
2018-03-06  178.250000  176.130005  177.910004  176.669998  23788500.0   
2018-03-07  175.850006  174.270004  174.940002  175.029999  31703500.0   
2018-03-08  177.119995  175.070007  175.479996  176.940002  23774100.0   
2018-03-09  180.000000  177.389999  177.960007  179.979996  32185200.0   
2018-03-10  180.000000  177.389999  177.960007  179.979996  32185200.0   

Attributes   Adj Close  
2018-03-01  171.713928  
2018-03-02  172.901230  
2018-03-03  172.901230  
2018-03-04  172.901230  
2018-03-05  173.499771  
2018-03-06  173.352600  
2018-03-07  171.743393  
2018-03-08  173.617508  
2018-03-09  176.600418  
2018-03-10  176.600418

# Group By

Reshaping data is an important component of any data wrangling toolkit as it allows the analyst to "massage" the data into the desired shape for further processing. 

Another equally important technique is the group by operation. Analysts having some experience with SQL or other data analysis toolsets (R's `tidyverse` for example) will find the group by operation a familiar strategy in many analysis-heavy workflow.

Consider the following DataFrame:

In [62]:
aapl = stock.xs('AAPL', level='Symbols', axis=1)
aapl = aapl.round(2)
aapl['Close_Diff'] = aapl['Close'].diff()
aapl['Weekday'] = aapl.index.weekday_name
aapl['Month'] = aapl.index.month_name()
aapl.tail()

Attributes    High     Low    Open   Close      Volume  Adj Close  Close_Diff  \
Date                                                                            
2019-04-17  203.38  198.61  199.54  203.13  28906800.0     202.35        3.88   
2019-04-18  204.15  202.52  203.12  203.86  24195800.0     203.08        0.73   
2019-04-22  204.94  202.34  202.83  204.53  19439500.0     203.75        0.67   
2019-04-23  207.75  203.90  204.43  207.48  23323000.0     206.68        2.95   
2019-04-24  208.48  207.05  207.36  207.16  17540600.0     206.37       -0.32   

Attributes    Weekday  Month  
Date                          
2019-04-17  Wednesday  April  
2019-04-18   Thursday  April  
2019-04-22     Monday  April  
2019-04-23    Tuesday  April  
2019-04-24  Wednesday  April

Pay special attention to how the `Close_Diff` column was created. It's the difference between the `Close` value of a stock price on a given day and the following day.

Supposed we want to compare the `Close_Diff` between each `Weekday`; On average, does Tuesday record a higher difference between the Close price of Apple stock compared to a Thursday?

In [63]:
aapl.groupby('Weekday').mean()

Attributes        High         Low        Open       Close        Volume  \
Weekday                                                                    
Friday      187.314848  183.676970  185.743636  185.425455  3.663884e+07   
Monday      187.599516  183.823065  185.788226  185.754516  3.276606e+07   
Thursday    187.768806  184.261940  185.905075  186.024179  3.297499e+07   
Tuesday     188.197910  184.630896  186.133284  186.539254  3.084036e+07   
Wednesday   188.078358  184.316418  186.125224  186.287612  3.232156e+07   

Attributes   Adj Close  Close_Diff  
Weekday                             
Friday      183.111212   -0.464848  
Monday      183.462903    0.000484  
Thursday    183.659403    0.084776  
Tuesday     184.172985    0.313485  
Wednesday   183.942388    0.584776

We can pass in multiple columns to the `.groupby()` call like so:

In [64]:
aapl.groupby(by=['Weekday', 'Month']).mean().head(15)

Attributes               High         Low        Open       Close  \
Weekday Month                                                       
Friday  April      180.185000  176.708333  179.333333  177.836667   
        August     216.312000  213.284000  214.188000  215.378000   
        December   165.062500  159.440000  164.212500  160.232500   
        February   169.937500  166.391250  168.485000  168.145000   
        January    165.321250  162.722500  163.827500  164.648750   
        July       191.972500  189.092500  190.817500  190.430000   
        June       189.152000  186.678000  188.320000  188.162000   
        March      180.962222  177.538889  179.157778  179.350000   
        May        187.942500  184.850000  185.790000  186.827500   
        November   194.312000  189.254000  192.166000  191.270000   
        October    223.185000  216.880000  220.585000  220.502500   
        September  224.852500  221.135000  223.292500  222.135000   
Monday  April      183.233333  180.016667  181.108889  182.016667   
        August     214.530000  211.552500  213.140000  212.835000   
        December   166.858000  162.068000  164.318000  164.586000   

Attributes               Volume   Adj Close  Close_Diff  
Weekday Month                                            
Friday  April      3.459397e+07  175.483333   -1.930000  
        August     3.106042e+07  212.750000    1.356000  
        December   5.525532e+07  158.937500   -4.430000  
        February   4.141038e+07  166.090000   -0.312500  
        January    3.419538e+07  162.030000    1.745000  
        July       1.867482e+07  187.575000   -0.200000  
        June       3.235132e+07  185.342000   -0.256000  
        March      3.399699e+07  177.502222    0.168889  
        May        2.954305e+07  183.852500    1.310000  
        November   4.515566e+07  189.582000   -4.420000  
        October    3.856388e+07  217.957500    0.937500  
        September  4.719880e+07  219.575000   -1.487500  
Monday  April      2.864951e+07  179.921111    1.060000  
        August     2.553228e+07  210.200000    0.520000  
        December   4.385782e+07  163.252000    0.684000

Before we going into some practice, let's look at some **group by** strategies when dealing with MultiIndex DataFrames. Consider the following DataFrame:

In [65]:
stock.stack().head(8)

Attributes                 High          Low         Open        Close  \
Date       Symbols                                                       
2018-01-02 AAPL      172.300003   169.259995   170.160004   172.259995   
           FB        181.580002   177.550003   177.679993   181.419998   
           GOOGL    1075.979980  1053.020020  1053.020020  1073.209961   
2018-01-03 AAPL      174.550003   171.960007   172.529999   172.229996   
           FB        184.779999   181.330002   181.880005   184.669998   
           GOOGL    1096.099976  1073.430054  1073.930054  1091.520020   
2018-01-04 AAPL      173.470001   172.080002   172.539993   173.029999   
           FB        186.210007   184.100006   184.899994   184.330002   

Attributes              Volume    Adj Close  
Date       Symbols                           
2018-01-02 AAPL     25555900.0   168.339050  
           FB       18151900.0   181.419998  
           GOOGL     1588300.0  1073.209961  
2018-01-03 AAPL     29517900.0   168.309738  
           FB       16886600.0   184.669998  
           GOOGL     1565900.0  1091.520020  
2018-01-04 AAPL     22434600.0   169.091522  
           FB       13880900.0   184.330002

On the Columns, there is 1 level. On the Index, there are 2 levels: `Date` and `Symbols` respectively. Supposed we want to groupby the second level of the axis (`Symbols`), we can pass in `axis=0` (referring to the row) and `level=1` (referring to the second level of index). 

The following groups it by the stock symbol and aggregate the values from the remaining levels using the `median`: 

In [66]:
stock.stack().groupby(level=1, axis=0).median()

Attributes         High          Low         Open        Close      Volume  \
Symbols                                                                      
AAPL         186.149994   182.559998   183.899994   184.160004  29855800.0   
FB           173.399994   168.160004   170.669998   171.160004  20750500.0   
GOOGL       1132.349976  1111.010010  1119.000000  1120.540039   1721100.0   

Attributes    Adj Close  
Symbols                  
AAPL         181.664673  
FB           171.160004  
GOOGL       1120.540039

# Learn-by-Building
## Graded Assignment 2
### Exploratory Data Analysis

1. Start from reading the data in using the following line of code:

```
pd.read_pickle('data_cache/stock')
```

2. Use `.diff()` to calculate the difference between a stock's adjusted closing price (`Adj Close`) and the following day. Recall that `.diff()` defaults to `periods=1`, so it calculates the difference between an element and the same element of the previous row.

3. Calculate the mean difference in `Adj Close` for each of `FB`, `GOOGL`, and `AAPL` stocks. Which have the highest average?

    - [ ] `AAPL`
    - [ ] `FB`
    - [ ] `GOOGL`


4. Start from reading the stock data by unpickling from `data_cache/stock`, then stack the DataFrame. From the stacked DataFrame, call `.groupby()` with the following parameters:
    - `axis=0` for the row
    - `level=1` for the innermost index (`Symbols` instead of `Date`)
    Chain a mean aggregation on the `groupby` operation. 

5. Using the output from (4), which has the lowest `Volume` on average? The volume is the number of shares that are transacted (buy/sell) during a given day.

    - [ ] `AAPL`
    - [ ] `FB`
    - [ ] `GOOGL`
    
6. Read `data_input/techcrunch.csv`, a dataset that stores fundraising rounds and amounts from startup companies of different categories around the US. Use `groupby` and perform a `median` aggregation. Which `category` raised the most amount in funding on average (use the `median`):

    - [ ] `mobile`
    - [ ] `cleantech`
    - [ ] `biotech`
    - [ ] `consulting`

7. Among all companies in San Francisco, which of the following are not among the top 5 most funded (highest `raisedAmt`) companies?
    - Hint: You may need to subset for companies where `city` is `San Francisco`, then group by `company` and `sum` on `raisedAmt` within each group. If it's helpful, you may consider chaining `sort_values()` and set `ascending=False` to get the DataFrame sorted by `raisedAmt` from largest to smallest
    
    - [ ] `OpenTable`
    - [ ] `Friendster`
    - [ ] `Facebook`
    - [ ] `Snapfish`

In [201]:
## Your code below


## -- Solution code